Fix shader uniform storage conversions and crash

bitsawer opened the pull request:
* Fixes https://github.com/godotengine/godot/issues/64051
* Fixes https://github.com/godotengine/godot/issues/74255
* Fixes https://github.com/godotengine/godot/issues/75033
* Fixes https://github.com/godotengine/godot/issues/76135
* Fixes https://github.com/godotengine/godot/issues/76157
* Fixes https://github.com/godotengine/godot/issues/75191
* Fixes https://github.com/godotengine/godot/issues/66619
* Fixes https://github.com/godotengine/godot/issues/63064
* Fixes https://github.com/godotengine/godot/issues/74680
* Fixes https://github.com/godotengine/godot/issues/64832
* Fixes https://github.com/godotengine/godot/issues/77970

Also fixes an unreported (?) crash that can happen when passing too little data in for certain types using `set_shader_parameter()`. It was caused by incorrect bounds checking which would read and write garbage memory or rarely crash if the read memory was not mapped in the process.

This PR also allows more types to be used with GDScript arrays when calling `set_shader_parameter()`. It relies on the normal Variant behavior to figure out the conversion. Previously, the only thing that worked worked for setting many shader type arrays was creating a raw PackedXXArray.

Array auto-conversion already works for many classes (Vector2, Vector3 etc.) that have their own PackedXXXArray. For example there is PackedVector3Array. But some classes like Vector4, Transform2D, Transform3D, Basis, Projection etc. don't have these and thus did not receive the Variant array auto-conversion goodies. This PR adds some of that conversion behavior without adding new PackedXXXArray's for each class.

I created a new file for the conversion functions that is shared between Vulkan and OpenGL renderer. There are other files in there that are shared like that, but let me know if I should put the functions somewhere else.

I'm not 100% sure if this PR breaks compatibility if user has previously saved some global shader values in the project settings. After a some quick testing I think not, but someone more familiar with that area might need to confirm.

**edit 2:**

I also updated some more uniform/global conversions, some of them haven't worked at all in a long time, but I guess nobody either used them or bothered to make a bug report. Also, here's a simple test project that can be used to test this functionality:

~~[UniformTest.zip](https://github.com/godotengine/godot/files/11221490/UniformTest.zip) (OLD test project)~~

~~[UniformTester.zip](https://github.com/godotengine/godot/files/11285405/UniformTester.zip)~~

[UniformTester3.zip](https://github.com/godotengine/godot/files/11592136/UniformTester3.zip) (NEWEST test project)

This test project was generated in master branch (d6dde819be6e02de133fe1b9f97c8fa6ce8668bb) and contains all incorrect global uniform and shader uniform types and data saved to main.tscn and project.godot. Opening and running the project will show a black box in the master, as the data is not converted properly. After applying this PR, the box should be green in editor (this test saved resource data conversion) and dark blue when run pressing F5 (this test runtime `Shader.set_shader_parameter()` conversion). The cube also moves once per second to test some x, y and z values. The cube also rotates when demo is run (not in editor mode).

NOTE: When using Compatiblity renderer, remove or comment out line `#define GLOBALS_CHECK` in uniforms.gdshader, seems like all global uniforms are not yet supported there (or they are buggy).

bitsawer committed:
Fix shader uniform storage conversions and crash

bitsawer commented on editor/shader_globals_editor.cpp:
```diff
@@ -198,7 +198,7 @@ class ShaderGlobalsEditorInterface : public Object {
 				} break;
 				case RS::GLOBAL_VAR_TYPE_MAT2: {
-					pinfo.type = Variant::PACKED_INT32_ARRAY;
+					pinfo.type = Variant::PACKED_FLOAT32_ARRAY;
```
Fixed IVEC4 and UVEC4 too and pushed changes. I also re-checked the other types and switches, but I didn't spot any more obviously wrong types.

clayjohn reviewed (dismissed):
I'm happy with the changes. 

it would be nice to have @Chaosus take a look and confirm that the array stuff is correct

LaneSun commented:
Thanks for the work, it helps a lot.

Chaosus reviewed (dismissed):
Seems fine for me, and the code looks much cleaner now.

bitsawer commented:
Fixed a few more types in `shader_language.cpp` and also added a compatibility Color conversion to `convert_array_std140()` for Color types which fixes https://github.com/godotengine/godot/issues/76157. Note that only the Vulkan renderer needs the srgb-to-linear conversion part. However, the real fix is probably to investigate why the PackedColorArray is serialized as a variant Array containing Color-variants instead of PackedColorArray (see https://github.com/godotengine/godot/issues/76157#issuecomment-1510822139). Still, adding that Color handling is also generally useful and also adds some backwards compatibility.

It's pretty clear by now that this PR breaks compatibility to some extent, but I think its worth it. Previously too many things were just broken in some way and there is no easy way to add backwards conversion compatibility everywhere. However, most issues should be pretty easy to fix as long as we tell users to where to look during out next release.

If we want to be super careful about maintaining compatibility, the only reliable way is to pretty much to add more conversions to Variant. For example, currently Quaternion or Plane does not automatically convert to Vector4 etc.

YuriSizov commented:
@clayjohn @Chaosus Can you take a look again?

> However, most issues should be pretty easy to fix as long as we tell users to where to look during the next release.

Can you provide a list of what the incompatible changes are and how do you propose to address them in user projects?

YuriSizov commented:
> Then again #75701 was a breaking change and that was done just to improve performace.

Ultimately, it wasn't, the old notification value was kept in the engine specifically to avoid breaking compat. The behavior changed though, that is true.

clayjohn commented:
I think most of the changes that could be considered breaking are also places where the current behaviour is totally broken. Like changing the type on the global shader vec4 makes sense because currently it is totally broken and leaks garbage

bitsawer commented:
Rebased and updated the PR. It should now be 100% compatible with old projects. I also updated the example project and its description in the first post, it tests uniform combination values and types that are affected by this PR (including saved global and Shader resource uniform data). Also tested bug report MRP projects, this PR seems to fix them all.

As a result of allowing much better type conversion with good backwards compatibility, I had to clean up the code a bit more. The templated conversion code is more complex than the current one, but it's either that or adding a few thousand lines of copy-paste Variant cross-type conversions. Changes also make it trivial to add support for new Variant or uniform types in the future and they should automatically receive full conversion support.

Previously, there were also several indexing errors in those uniform conversion loops as mentioned before (BVECX heap over-reads, wrong loop step counter in UVEC4 resulting in reading heap garbage etc.). I also made a few minor empty line additions and removals to make the two `material_storage.cpp`-files look more similar and consistent, easier to compare side-by-side.

Also noticed and fixed a subtle (unreported?) bug between Vulkan and OpengGL storage, for example passing Vector2/3/4(i), Quaternion or Plane would sometimes lead to different or bad values in vec4 uniforms. The `material_storage.cpp` files had gotten out-of-sync and the order of type checks in vec4 branch was different, which caused several incorrect type conversion combinations in OpenGL compatibility renderer.

Also fixes https://github.com/godotengine/godot/issues/75191 (added to first post). The vec4 values remain the same in resource file, but now in the editor there is a revert button which will fix the vec4 values to correct defaults (as the editor notices that old incorrect default value is not currently correct).

Also fixes https://github.com/godotengine/godot/issues/66619 (added to first post).

While I tested this a lot, there might still be some subtle bugs around. Still, this PR should improve the current situation significantly.

bitsawer commented:
Also fixes https://github.com/godotengine/godot/issues/63064 and https://github.com/godotengine/godot/issues/74680 (added to first post). I made (and reverted) a separate PR for those (https://github.com/godotengine/godot/pull/76438), but as some uniforms have a wrong type currently that fix should be incorporated into this PR for better backwards compatibility so that all type updates happen at the same commit.

I'll need to test and fix a few small type compatibility issues so that all this stuff works together. Converted to draft until then.

bitsawer commented:
Finally had some time to go through all issues, MRP projects and demos to verify everything works together. I added a few compatibility tweaks and rebased this PR, it should be fully backwards compatible and I didn't notice any regressions during my testing. Still, there are a lot of changes so something might have slipped past me, but any issues should be easily fixable once someone reports them.

Ready for a review. Also updated the test project in the first post.

clayjohn commented on drivers/gles3/storage/material_storage.cpp:
```diff
@@ -39,32 +39,24 @@
-					gui[j + 1] = 0; // ignored
-					gui[j + 2] = 0; // ignored
-					gui[j + 3] = 0; // ignored
+				PackedInt32Array ba = value;
+				for (int i = 0; i < ba.size(); i++) {
+					ba.set(i, ba[i] ? 1 : 0);
```
Is this necessary? The code seems highly suspicious. It looks like it is copying over the entire ``value`` array, then iterating over it to clean the values, then passing the whole array to a new function to copy each value one-by-one into another array. Could we not convert, clean, and copy all at once as we do now?

Its nice that this templated approach cuts the amount of code in half, but I have a feeling it is forcing us to do way more copies and way more operations to do the same amount of work.

bitsawer replied:
Note that old code does convert+copy+iterate passes too, they are just harder to see. In the old way, these lines:

```cpp
const PackedInt32Array &ba = value;
```

create a new temporary: a PackedInt32Array from a value variant object using Variant conversion. It just binds it as a const reference, but a temporary is still made and allocated. So a full conversion and array allocation is done here, anyway, it's just harder to see. The following manual for-loop is replaced by `write_array_std140(`), but they do the same thing. The amount of work done should be roughly the same between the old and new code. And when passed the same values, the new code does not really do any more conversions than the old one, there is a fast path in `convert_array_std140()` for many packed arrays. The only extra copying comes when passing more complex object arrays which have to allocate some temporary extra space, but these didn't work at all in the old code. I suspect because doing these conversions for arbitrary arrays and objects properly is actually pretty hard, as evident. I though about special casing all types and use internal variant accessors in variant_internal.h to handle these conversions without temporary copies, but it turned out that kind of code is even harder to write, read and understand.

The only exception here is indeed the boolean uniform handling (bool, bvec2, bvec3, bvec4), which has to do extra work for the "force to 1 or 0"-loop. However, the only extra cost is the loop (which is typically much less than 100 elements), because the temporary COW packed array is already made and its CowData reference count is 1, which means it doesn't have to copy anything when written into because it is already a non-shared copy which had to be made anyway (both in old and new versions).

In practice, it shouldn't have much effect on performance, but uniform boolean conversions are probably is a bit slower than the original version for the same arguments. Of course, I doubt many people actually use uniform boolean in the first place. However, to match the original performance and not make copy of the COW vector, I could write:

```cpp
const PackedInt32Array &ba = convert_array_std140<Vector2i, int32_t>(value);
write_array_std140<Vector2i>(ba, gui, p_array_size, 4);
for (int = 0; i < p_array_size * 4) {
    gui[i] = gui[i] ? 1 : 0;
}
```

It still makes one extra loop over the data, but as uniform arrays are not usually very large in practice it shouldn't matter too much. Alternatively, we could add an optional lambda function converter to inline the conversion into the copy loop, which the compiler can optimize nicely:

```cpp
const PackedInt32Array &ba = convert_array_std140<Vector2i, int32_t>(value);
write_array_std140<Vector2i>(ba, gui, p_array_size, 4, [](int32_t value) { return value ? 1: 0 });
```

With this version, the performance should be pretty much identical. Or we could leave the new code as-is, it's pretty simple to understand this way with a pretty minimal performance penalty.

clayjohn replied:
Thank you for the detailed response. I didn't realize that ``const PackedInt32Array &ba = value;`` was copying the entire array.

Indeed the boolean arrays are the worst offenders in terms of copying necessarily, but they are also the cheapest to copy and likely won't be that large. 

I'm happy with the current state of things in that case. I would avoid making it any more complicated. Already this PR adds significant complexity by adding layers of templated functions. I understand that was done in order to make handling all the edge cases manageable, but adding more complexity will make this impossible to debug six months from now.

clayjohn reviewed (approved):
Seeing as my only concern turned out to be my misunderstanding, I am happy to go ahead with this PR. I trust that @bitsawer has properly handled things and this is a significant improvement over the status quo. 

I would go ahead and merge this so we can get wide testing before releasing 4.1

akien-mga commented:
Thanks a ton!
